File: 2022/java/Day6.java
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.stream.IntStream;

class Day6 {
    public static void main(String ... args) throws Exception {
        var input = Files.readString(Path.of("../input/day6.txt"));
        System.out.printf("part 1: %d\n", findMarker(input, 4));
        System.out.printf("part 2: %d\n", findMarker(input, 14));
    }

    static int findMarker(String input, int len) {
        return IntStream.range(0, input.length()-len)
            .filter(i -> allDifferent(input.substring(i, i + len))).findFirst().getAsInt() + len;
    }

    static boolean allDifferent(String input) {
        return input.chars().distinct().count() == input.length();
    }
}

File: 2022/go/day6/day6.go
package main

import (
	"fmt"
	"os"
)

func main() {
	input, _ := os.ReadFile("../../input/day6.txt")
	fmt.Printf("part1: %d\n", findMarker(input, 4))
	fmt.Printf("part2: %d\n", findMarker(input, 14))
}

func findMarker(input []byte, len int) int {
	for i := range input {
		if allDifferent(input[i : i+len]) {
			return i + len
		}
	}
	return -1
}

func allDifferent(s []byte) bool {
	bytes := map[byte]bool{}
	for _, c1 := range s {
		bytes[c1] = true
	}
	return len(bytes) == len(s)
}
